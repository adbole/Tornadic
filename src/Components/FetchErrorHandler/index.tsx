import React from "react";
import { SWRConfig, useSWRConfig } from "swr";


export default function FetchErrorHandler({
    children,
    errorRender,
}: {
    children: React.ReactNode;
    errorRender: (hasError: boolean, retry: () => void) => React.ReactNode;
}) {
    const { mutate } = useSWRConfig();
    const [errorKeys, setErrorKeys] = React.useState<Set<string>>(new Set);

    return (
        <>
            {errorRender(errorKeys.size > 0, () => {
                const keys = errorKeys;
                setErrorKeys(new Set);

                keys.forEach(key => mutate(key));
            })}
            <SWRConfig
                value={{
                    onError(_, key) {
                        setErrorKeys(prev => new Set([...prev, key]));
                    },
                }}
            >
                {children}
            </SWRConfig>
        </>
    );
}
