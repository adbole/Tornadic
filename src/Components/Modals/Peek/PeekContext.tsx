import React from "react";

import { useOpenMeteo } from "Hooks";

import { throwError } from "ts/Helpers";
import type NWSAlert from "ts/NWSAlert";
import type Weather from "ts/Weather";


const Context = React.createContext<{
    weather: Weather;
    alerts: NWSAlert[];
}>({} as any);

export const usePeekWeather = () =>
    React.useContext(Context) || throwError("Please use usePeekWeather within a PeekContext");

export default function PeekContext({
    latitude,
    longitude,
    children,
}: {
    latitude?: number;
    longitude?: number;
    children: React.ReactNode;
}) {
    const { weather, alerts } = useOpenMeteo(latitude, longitude);

    const value = React.useMemo(() => {
        if (!weather || !alerts) return null;

        return {
            weather,
            alerts,
        };
    }, [weather, alerts]);

    if (!value) return (
        <div>
            <span className="text-loader"/>
            <span className="text-loader"/>
            <span className="text-loader"/>
            <span className="text-loader"/>
        </div>
    );

    return <Context.Provider value={value}>{children}</Context.Provider>;
}
