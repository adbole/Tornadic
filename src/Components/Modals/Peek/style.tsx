import styled from "@emotion/styled";

import DailyWidget from "Components/Daily/style";
import HourlyWidget from "Components/Hourly/style";

import { varNames, vars } from "ts/StyleMixins";

import { ModalContent } from "../Modal";


export default styled(ModalContent)({
    padding: "10px",

    display: "grid",
    gridTemplateColumns: "repeat(2, 1fr)",
    gridAutoRows: "150px",
    gap: "10px",

    [`${HourlyWidget}`] : {
        gridColumn: "span 2"
    },

    [`${DailyWidget}`] : {
        gridArea: "unset",
        gridColumn: "span 2",
        gridRow: "span 2"
    }
});

export const ErrorMessage = styled.div({
    display: "flex",
    flexDirection: "column",
    justifyContent: "center",
    alignItems: "center",

    borderRadius: vars.borderRadius,

    background: vars.warn,
    color: "white",
    padding: "10px",
    gridColumn: " 1 / -1",

    [varNames.svgSize]: "2rem",
    svg: { fill: "white" },
});
