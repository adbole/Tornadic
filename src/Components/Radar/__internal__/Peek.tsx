import React from "react";
import { Popup, useMap } from "react-leaflet";
import { Global } from "@emotion/react";
import styled from "@emotion/styled";
import type L from "leaflet";

import { useNullableState } from "Hooks";

import PeekModal from "Components/Modals/Peek";

import { vars } from "ts/StyleMixins";


const HoldPopup = styled(Popup)({
    marginBottom: 0,
    width: "100px",
    height: "100px",
    borderRadius: vars.borderRadius,
    backdropFilter: "blur(5px)",
    border: "1px solid white",
    pointerEvents: "none",
    touchAction: "none",
    "> .leaflet-popup-content-wrapper": { display: "none" },
    transition: "opacity 0.9s linear !important",
});

/**
 * A modal displaying some weather information for the given location
 */
export default function Peek() {
    const map = useMap();
    const [position, setPosition, unsetPosition] = useNullableState<L.LatLng>();
    const [latlng, setLatLng, unsetLatLng] = useNullableState<L.LatLng>();

    const timeout = React.useRef<NodeJS.Timeout>();
    const initialXY = React.useRef<[number, number]>([0, 0]);

    const beginHold = React.useCallback(
        (pos: L.LatLng, xy: [number, number]) => {
            if (!map.dragging.enabled()) return;

            if (timeout.current) {
                clearTimeout(timeout.current);
                timeout.current = undefined;
                unsetPosition();

                return;
            }

            //100ms delay allows for a normal click or tap to pass through without issues.
            timeout.current = setTimeout(() => {
                setPosition(pos);
                initialXY.current = xy;

                timeout.current = setTimeout(() => {
                    unsetPosition();
                    setLatLng(pos);
                }, 900);
            }, 100);
        },
        // eslint-disable-next-line react-hooks/exhaustive-deps
        [map]
    );

    const onMouseDown = React.useCallback(
        ({ latlng, originalEvent: { clientX, clientY } }: L.LeafletMouseEvent) =>
            beginHold(latlng, [clientX, clientY]),
        [beginHold]
    );

    const onTouchStart = React.useCallback(
        ({ touches: [touch] }: TouchEvent) => {
            const latlng = map.containerPointToLatLng([touch.clientX, touch.clientY]);

            beginHold(latlng, [touch.clientX, touch.clientY]);
        },
        [map, beginHold]
    );

    const onUp = React.useCallback(() => {
        if (!timeout.current) return;

        clearTimeout(timeout.current);
        timeout.current = undefined;
        unsetPosition();
    }, [unsetPosition]);

    const onMove = React.useCallback(
        (e: L.LeafletMouseEvent | TouchEvent) => {
            if (!timeout.current) return;

            const event =
                window.TouchEvent && e instanceof TouchEvent
                    ? e.touches[e.touches.length - 1]
                    : (e as L.LeafletMouseEvent).originalEvent;

            const xChange = Math.abs(event.clientX - initialXY.current[0]);
            const yChange = Math.abs(event.clientY - initialXY.current[1]);
            const maxChange = 20;

            if (xChange > maxChange || yChange > maxChange) {
                onUp();
            }
        },
        [onUp]
    );

    React.useEffect(() => {
        const mapEvents = [
            ["mousedown", onMouseDown],
            ["mousemove", onMove],
            ["mouseup", onUp],
            ["contextmenu", () => undefined],
        ] as const;

        const containerEvents = [
            ["touchstart", onTouchStart],
            ["touchend", onUp],
            ["touchmove", onMove],
        ] as const;

        mapEvents.forEach(([event, handler]) => map.on(event, handler));

        const mapContainer = map.getContainer();
        containerEvents.forEach(([event, handler]) =>
            mapContainer.addEventListener(event, handler, { passive: true })
        );

        return () => {
            mapEvents.forEach(([event, handler]) => map.off(event, handler));
            containerEvents.forEach(([event, handler]) =>
                mapContainer.removeEventListener(event, handler)
            );
        };
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [map, beginHold]);

    return (
        <>
            {position && (
                <>
                    <HoldPopup position={position} closeButton={false} offset={[0, 50]} />
                    <Global styles={{ body: { userSelect: "none" } }} />
                </>
            )}
            <PeekModal
                latitude={latlng?.lat}
                longitude={latlng?.lng}
                isOpen={Boolean(latlng)}
                onClose={unsetLatLng}
            />
        </>
    );
}
