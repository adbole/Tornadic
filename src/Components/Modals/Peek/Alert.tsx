import { useBooleanState } from "Hooks";

import AlertModal from "Components/Alert/AlertModal";
import { AlertInformationDisplay } from "Components/Alert/Common";

import { usePeekWeather } from "./PeekContext";


export default function Alert() {
    const { alerts } = usePeekWeather()
    const [modalOpen, showModal, hideModal] = useBooleanState(false);

    //If no alerts are active then don't display this component.
    if (!alerts.length) return <></>;

    //Determine which alert should be shown.
    const alertToShow = alerts.reduce(
        (highest, next) => (next.priority < highest.priority ? next : highest),
        alerts[0]
    );

    return (
        <>
            <div className={`peek-alert ${alertToShow.getAlertCSS()}`} onClick={showModal}>
                <div>
                    <AlertInformationDisplay alert={alertToShow} />
                </div>

                {alerts.length > 1 && (
                    <p className="excess-alerts">+{alerts.length - 1} more alert(s)</p>
                )}
            </div>
            <AlertModal alerts={alerts} isOpen={modalOpen} onClose={hideModal} />
        </>
    );
}
