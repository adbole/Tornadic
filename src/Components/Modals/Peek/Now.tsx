import { usePeekWeather } from "./PeekContext"


export default function Now() {
    const { weather } = usePeekWeather()

    const {
        temperature,
        feelsLike,
        conditionInfo
    } = weather.getNow()

    return (
        <div className={`peek-now ${conditionInfo.background}`}>
            <p>{weather.point.properties.relativeLocation.properties.city}</p>
            <h1>{temperature}</h1>
            <p>{conditionInfo.intensity} {conditionInfo.type}</p>
            <p>Feels like {feelsLike}</p>
        </div>
    )
}