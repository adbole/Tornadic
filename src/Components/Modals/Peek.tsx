import React from "react";
import styled from "@emotion/styled";

import WeatherContext from "Contexts/WeatherContext";

import Alert from "Components/Alert";
import GaugeWidget from "Components/GagueWidget";
import HazardLevel from "Components/HazardLevel";
import { Button } from "Components/Input";
import Now from "Components/Now";
import Pressure from "Components/Pressure";
import SimpleInfoWidget from "Components/SimpleInfoWidget";
import SkeletonWidget from "Components/SkeletonWidget";
import { WidgetStyle } from "Components/Widget";
import Wind from "Components/Wind";
import { ExclamationTriangle } from "svgs";
import { Droplet, Eye, Moisture, Thermometer } from "svgs/widget";

import type { ModalProps } from "./Modal";
import Modal, { ModalContent } from "./Modal";


const PeekContent = styled(ModalContent)({
    padding: 0,

    display: "grid",
    gridTemplateColumns: "repeat(4, 1fr)",
    gridAutoRows: "auto",

    [`${WidgetStyle}`]: {
        backdropFilter: "none",
        boxShadow: "none",
        minHeight: "150px",
    },

    [`${Now.Style}, ${Alert.Style}`]: {
        gridColumn: "span 4",
        borderRadius: 0,
    },

    [`${WidgetStyle}:nth-of-type(n+ 7)`]: { gridColumn: "span 2" },
});

export default function Peek({
    latitude,
    longitude,
    ...modalProps
}: {
    latitude?: number;
    longitude?: number;
} & ModalProps) {
    return (
        <Modal {...modalProps}>
            <PeekContent>
                <WeatherContext
                    latitude={latitude}
                    longitude={longitude}
                    skeletonRender={() => (
                        <>
                            <SkeletonWidget className="now" size="widget-large" />
                            <SkeletonWidget className="alert" size="widget-wide" />

                            {Array.from({ length: 4 }, (_, i) => (
                                <SkeletonWidget key={i} />
                            ))}

                            {Array.from({ length: 4 }, (_, i) => (
                                <SkeletonWidget className="wind" key={i} />
                            ))}
                        </>
                    )}
                    fallbackRender={getData => (
                        <>
                            <ExclamationTriangle />
                            <p>Unable to get weater data</p>
                            <Button onClick={getData}>Try Again</Button>
                        </>
                    )}
                >
                    <Now displayOnly />
                    <Alert />
                    <SimpleInfoWidget
                        icon={<Droplet />}
                        title="Precipitation"
                        property="precipitation"
                    />
                    <SimpleInfoWidget
                        icon={<Thermometer />}
                        title="Dewpoint"
                        property="dewpoint_2m"
                    />
                    <SimpleInfoWidget
                        icon={<Moisture />}
                        title="Humidity"
                        property="relativehumidity_2m"
                    />
                    <SimpleInfoWidget icon={<Eye />} title="Visibility" property="visibility" />

                    <HazardLevel hazard="us_aqi" />
                    <HazardLevel hazard="uv_index" />

                    <Wind />
                    <Pressure />
                </WeatherContext>
            </PeekContent>
        </Modal>
    );
}
