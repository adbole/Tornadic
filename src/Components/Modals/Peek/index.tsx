import WeatherContext from "Contexts/WeatherContext";

import Alert from "Components/Alert";
import HazardLevel from "Components/HazardLevel";
import { Button } from "Components/Input";
import Now from "Components/Now";
import Pressure from "Components/Pressure";
import SimpleInfoWidget from "Components/Simple";
import Skeleton from "Components/Skeleton";
import Wind from "Components/Wind";
import { ExclamationTriangle } from "svgs";
import { Droplet, Eye, Moisture, Thermometer } from "svgs/widget";

import type { ModalProps } from "../Modal";
import Modal from "../Modal";

import PeekContent, { AlertSkeleton,NowSkeleton } from "./style";


export default function Peek({
    latitude,
    longitude,
    ...modalProps
}: {
    latitude?: number;
    longitude?: number;
} & ModalProps) {
    return (
        <Modal {...modalProps}>
            <PeekContent>
                <WeatherContext
                    latitude={latitude}
                    longitude={longitude}
                    skeletonRender={() => (
                        <>
                            <NowSkeleton className="now" size="widget-large" />
                            <AlertSkeleton className="alert" />

                            {Array.from({ length: 4 }, (_, i) => (
                                <Skeleton key={i} />
                            ))}

                            {Array.from({ length: 4 }, (_, i) => (
                                <Skeleton key={i} />
                            ))}
                        </>
                    )}
                    fallbackRender={getData => (
                        <>
                            <ExclamationTriangle />
                            <p>Unable to get weater data</p>
                            <Button onClick={getData}>Try Again</Button>
                        </>
                    )}
                >
                    <Now className="now" displayOnly />
                    <Alert className="alert" />
                    <SimpleInfoWidget
                        icon={<Droplet />}
                        title="Precipitation"
                        property="precipitation"
                    />
                    <SimpleInfoWidget
                        icon={<Thermometer />}
                        title="Dewpoint"
                        property="dewpoint_2m"
                    />
                    <SimpleInfoWidget
                        icon={<Moisture />}
                        title="Humidity"
                        property="relativehumidity_2m"
                    />
                    <SimpleInfoWidget icon={<Eye />} title="Visibility" property="visibility" />

                    <HazardLevel hazard="us_aqi" />
                    <HazardLevel hazard="uv_index" />

                    <Wind />
                    <Pressure />
                </WeatherContext>
            </PeekContent>
        </Modal>
    );
}
