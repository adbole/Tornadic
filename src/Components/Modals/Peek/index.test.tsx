import { act, render, screen } from "@testing-library/react";

import FetchErrorHandler from "Components/FetchErrorHandler";

import Peek from ".";

//Peek acts as a mini version of the application as a whole, so tests here
//are only checking that the correct components are rendered and get proper props
//hence all the mocks

const mockFetchErrorHanlder = "FetchErrorHandler";
const mockRetry = vi.fn();
const mockWeatherContext = "WeatherContext";
const mockNow = "Now";
const mockAlert = "Alert";
const mockBackground = "Background";
const mockHourly = "Hourly";
const mockDaily = "Daily";
const mockWind = "Wind";
const mockPressure = "Pressure";
const mockHazard = "HazardLevel";
const mockSunTime = "SunTime";

vi.mock("Components/FetchErrorHandler");

// prettier-ignore
const mockFetchErrorHandlerFactory = (error: boolean) => (
    // eslint-disable-next-line react/function-component-definition
    ({ errorRender, children }: { 
        errorRender: (hasError: boolean, retry: () => void) => JSX.Element, 
        children: React.ReactNode 
    }) => (
        <div data-testid={mockFetchErrorHanlder}>
            {errorRender(error, mockRetry)}
            {children}
        </div>
    ) 
)

vi.mock("Contexts/WeatherContext", () => ({
    default: ({
        latitude,
        longitude,
        children,
        skeletonRender,
    }: {
        latitude: number;
        longitude: number;
        children: React.ReactNode;
        skeletonRender: () => JSX.Element;
    }) => (
        <div data-testid={mockWeatherContext}>
            <p>Lat: {latitude}</p>
            <p>Long: {longitude}</p>
            {skeletonRender()}
            {children}
        </div>
    ),
}));

vi.mock("Components/Now", () => ({
    default: ({ displayOnly }: { displayOnly: boolean }) => (
        <div data-testid={mockNow}>
            <p>Display Only: {displayOnly.toString()}</p>
        </div>
    ),
}));

vi.mock("Components/Alert", () => ({ default: () => <div data-testid={mockAlert} /> }));
vi.mock("Components/Background", () => ({ default: () => <div data-testid={mockBackground} /> }));
vi.mock("Components/Hourly", () => ({ default: () => <div data-testid={mockHourly} /> }));
vi.mock("Components/Daily", () => ({ default: () => <div data-testid={mockDaily} /> }));

vi.mock("Components/Simple", () => ({
    default: ({
        icon,
        title,
        property,
    }: {
        icon: React.ReactNode;
        title: string;
        property: string;
    }) => (
        <div>
            <p>Icon: {icon}</p>
            <p>Title: {title}</p>
            <p>Property: {property}</p>
        </div>
    ),
}));

vi.mock("Components/Wind", () => ({ default: () => <div data-testid={mockWind} /> }));
vi.mock("Components/Pressure", () => ({ default: () => <div data-testid={mockPressure} /> }));
vi.mock("Components/SunTime", () => ({ default: () => <div data-testid={mockSunTime} /> }));

vi.mock("Components/HazardLevel", () => ({
    default: ({ hazard }: { hazard: string }) => (
        <div data-testid={mockHazard}>
            <p>Hazard: {hazard}</p>
        </div>
    ),
}));

vi.mock("svgs/widget", () => ({
    Droplet: () => <span>Droplet</span>,
    Eye: () => <span>Eye</span>,
    Moisture: () => <span>Moisture</span>,
    Thermometer: () => <span>Thermometer</span>,
}));

test("Renders nothing if isOpen is false", () => {
    render(<Peek isOpen={false} onClose={vi.fn()} />);

    expect(screen.queryByRole("dialog")).not.toBeInTheDocument();
});

test("Renders a dialog if isOpen is true", () => {
    render(<Peek isOpen={true} onClose={vi.fn()} />);

    expect.soft(screen.queryByRole("dialog")).toBeInTheDocument();
});

describe("FetchErrorHandler", () => {
    test("Renders the FetchErrorHandler with with the error shown", () => {
        vi.mocked(FetchErrorHandler).mockImplementation(mockFetchErrorHandlerFactory(true));

        render(<Peek isOpen={true} onClose={vi.fn()} />);

        expect.soft(screen.queryByTestId(mockFetchErrorHanlder)).toBeInTheDocument();
        expect.soft(screen.queryByText("Couldn't get weather info")).toBeInTheDocument();

        const retryBtn = screen.getByText("Try Again");

        expect.soft(retryBtn).toBeInTheDocument();

        act(() => {
            retryBtn.click();
        });

        expect.soft(mockRetry).toHaveBeenCalledOnce();
    });

    test("Error has display: None if FetchErrorHandler passes false to errorRender", () => {
        vi.mocked(FetchErrorHandler).mockImplementation(mockFetchErrorHandlerFactory(false));

        render(<Peek isOpen={true} onClose={vi.fn()} />);

        expect.soft(screen.queryByTestId(mockFetchErrorHanlder)).toBeInTheDocument();

        expect
            .soft(screen.getByText("Couldn't get weather info").parentElement)
            .toHaveStyle({ display: "none" });
    });
});

test("Matches Snapshot", () => {
    render(<Peek isOpen={true} onClose={vi.fn()} latitude={365} longitude={405} />);

    expect(document.body).toMatchSnapshot();
})