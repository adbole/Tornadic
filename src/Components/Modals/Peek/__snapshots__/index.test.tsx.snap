// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Matches Snapshot 1`] = `
@keyframes animation-0 {
  to {
    background-position-x: -200%;
  }
}

.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow: hidden;
  margin: auto;
  padding: 0;
  width: 500px;
  color: inherit;
  background-color: var(--background);
  border-radius: var(--border-radius);
  border: none;
  outline: none;
  -webkit-transform: translateY(100vh);
  -moz-transform: translateY(100vh);
  -ms-transform: translateY(100vh);
  transform: translateY(100vh);
  -webkit-transition: -webkit-transform 1s ease;
  transition: transform 1s ease;
  -webkit-transform: translateY(0)!important;
  -moz-transform: translateY(0)!important;
  -ms-transform: translateY(0)!important;
  transform: translateY(0)!important;
}

.emotion-0::backdrop {
  -webkit-backdrop-filter: blur(5px);
  backdrop-filter: blur(5px);
  background-color: rgba(0, 0, 0, 0.5);
  opacity: 0;
  -webkit-transition: opacity 1s ease;
  transition: opacity 1s ease;
  cursor: pointer;
}

@media screen and (max-width: 800px) {
  .emotion-0 {
    max-width: 100%;
    max-height: 75%;
    margin: auto auto 0px auto;
    border-radius: var(--border-radius) var(--border-radius) 0px 0px;
    -webkit-transform: translateY(100%);
    -moz-transform: translateY(100%);
    -ms-transform: translateY(100%);
    transform: translateY(100%);
  }
}

.emotion-0::backdrop {
  opacity: 1;
}

.emotion-2 {
  overflow-y: auto;
  padding: 20px;
  padding: 10px;
  display: grid;
  grid-template-columns: repeat(2, 1fr);
  grid-auto-rows: 150px;
  gap: 10px;
}

.emotion-2 h2,
.emotion-2 h3,
.emotion-2 h4,
.emotion-2 h5,
.emotion-2 h6 {
  font-weight: 500;
}

.emotion-2 hr {
  height: 1px;
  background-color: #efdddd57;
  border: none;
  margin: 5px;
}

.emotion-2>p {
  margin-bottom: 10px;
}

.emotion-2 .e7hfcrb3 {
  grid-column: span 2;
}

.emotion-2 .e1xe5sf55 {
  grid-area: unset;
  grid-column: span 2;
  grid-row: span 2;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: var(--border-radius);
  background: var(--warn);
  color: white;
  padding: 10px;
  grid-column: 1/-1;
  --svg-size: 2rem;
}

.emotion-4 svg {
  fill: white;
}

.emotion-6 {
  -webkit-transition: filter 0.3s ease;
  transition: filter 0.3s ease;
  border: none;
  border-radius: var(--input-border-radius);
  font-size: inherit;
  box-shadow: 0px 4px 10px rgba(0, 0, 0, 0.25);
  margin: 5px 0px;
  color: inherit;
  background-color: var(--primary);
  padding: 5px 15px;
}

.emotion-6:hover,
.emotion-6:focus {
  cursor: pointer;
}

.emotion-6:active {
  -webkit-filter: brightness(80%);
  filter: brightness(80%);
}

.emotion-6:disabled {
  cursor: not-allowed;
  -webkit-filter: grayscale(0.5);
  filter: grayscale(0.5);
}

.emotion-9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: relative;
  padding: 10px;
  overflow: hidden;
  -webkit-backdrop-filter: blur(10px);
  backdrop-filter: blur(10px);
  box-shadow: inset 0 0 0 200px rgba(255, 255, 255, 0.01);
  border: 1px solid rgba(255, 255, 255, 0.1);
  border-radius: var(--border-radius);
  min-height: 300px;
  grid-column: span 2;
  grid-row: span 2;
  background: linear-gradient(90deg, rgba(255, 255, 255, 0) 0, rgba(255, 255, 255, 0.1) 20%, rgba(255, 255, 255, 0.15) 60%, rgba(255, 255, 255, 0));
  -webkit-background-size: 200%,100%;
  background-size: 200%,100%;
  -webkit-animation: animation-0 1.5s linear infinite;
  animation: animation-0 1.5s linear infinite;
}

.emotion-9>:not(.e5hllwo1) {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.emotion-12 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: relative;
  padding: 10px;
  overflow: hidden;
  -webkit-backdrop-filter: blur(10px);
  backdrop-filter: blur(10px);
  box-shadow: inset 0 0 0 200px rgba(255, 255, 255, 0.01);
  border: 1px solid rgba(255, 255, 255, 0.1);
  border-radius: var(--border-radius);
  grid-column: span 2;
  background: linear-gradient(90deg, rgba(255, 255, 255, 0) 0, rgba(255, 255, 255, 0.1) 20%, rgba(255, 255, 255, 0.15) 60%, rgba(255, 255, 255, 0));
  -webkit-background-size: 200%,100%;
  background-size: 200%,100%;
  -webkit-animation: animation-0 1.5s linear infinite;
  animation: animation-0 1.5s linear infinite;
}

.emotion-12>:not(.e5hllwo1) {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.emotion-17 {
  display: grid;
  gap: inherit;
}

.emotion-20 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: relative;
  padding: 10px;
  overflow: hidden;
  -webkit-backdrop-filter: blur(10px);
  backdrop-filter: blur(10px);
  box-shadow: inset 0 0 0 200px rgba(255, 255, 255, 0.01);
  border: 1px solid rgba(255, 255, 255, 0.1);
  border-radius: var(--border-radius);
  background: linear-gradient(90deg, rgba(255, 255, 255, 0) 0, rgba(255, 255, 255, 0.1) 20%, rgba(255, 255, 255, 0.15) 60%, rgba(255, 255, 255, 0));
  -webkit-background-size: 200%,100%;
  background-size: 200%,100%;
  -webkit-animation: animation-0 1.5s linear infinite;
  animation: animation-0 1.5s linear infinite;
}

.emotion-20>:not(.e5hllwo1) {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

<body>
  <div />
  <dialog
    class="emotion-0 emotion-1"
    open=""
  >
    <div
      class="emotion-2 emotion-3"
    >
      <div
        data-testid="FetchErrorHandler"
      >
        <div
          class="emotion-4 emotion-5"
          style="display: none;"
        >
          <svg
            fill="red"
            viewBox="0 0 16 16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8.982 1.566a1.13 1.13 0 0 0-1.96 0L.165 13.233c-.457.778.091 1.767.98 1.767h13.713c.889 0 1.438-.99.98-1.767L8.982 1.566zM8 5c.535 0 .954.462.9.995l-.35 3.507a.552.552 0 0 1-1.1 0L7.1 5.995A.905.905 0 0 1 8 5zm.002 6a1 1 0 1 1 0 2 1 1 0 0 1 0-2z"
            />
          </svg>
          <span>
            An error occured while getting the data
          </span>
          <button
            class="emotion-6 emotion-7"
            type="button"
          >
            Try Again Now
          </button>
        </div>
        <div
          data-testid="WeatherContext"
        >
          <p>
            Lat: 
            365
          </p>
          <p>
            Long: 
            405
          </p>
          <section
            class=" emotion-8 emotion-9 emotion-10"
            data-testid="WIDGET_SECTION"
          />
          <section
            class=" emotion-8 emotion-12 emotion-10"
            data-testid="WIDGET_SECTION"
          />
          <section
            class=" emotion-8 emotion-9 emotion-10"
            data-testid="WIDGET_SECTION"
          />
          <div
            class="emotion-17 emotion-18"
          >
            <section
              class=" emotion-8 emotion-20 emotion-10"
              data-testid="WIDGET_SECTION"
            />
            <section
              class=" emotion-8 emotion-20 emotion-10"
              data-testid="WIDGET_SECTION"
            />
          </div>
          <div
            class="emotion-17 emotion-18"
          >
            <section
              class=" emotion-8 emotion-20 emotion-10"
              data-testid="WIDGET_SECTION"
            />
            <section
              class=" emotion-8 emotion-20 emotion-10"
              data-testid="WIDGET_SECTION"
            />
          </div>
          <section
            class=" emotion-8 emotion-20 emotion-10"
            data-testid="WIDGET_SECTION"
          />
          <section
            class=" emotion-8 emotion-20 emotion-10"
            data-testid="WIDGET_SECTION"
          />
          <section
            class=" emotion-8 emotion-20 emotion-10"
            data-testid="WIDGET_SECTION"
          />
          <section
            class=" emotion-8 emotion-20 emotion-10"
            data-testid="WIDGET_SECTION"
          />
          <section
            class=" emotion-8 emotion-12 emotion-10"
            data-testid="WIDGET_SECTION"
          />
          <div
            data-testid="Now"
          >
            <p>
              Display Only: 
              true
            </p>
          </div>
          <div
            data-testid="Background"
          />
          <div
            data-testid="Alert"
          />
          <div
            data-testid="Hourly"
          />
          <div
            data-testid="Daily"
          />
          <div
            class="emotion-17 emotion-18"
          >
            <div>
              <p>
                Icon: 
                <span>
                  Droplet
                </span>
              </p>
              <p>
                Title: 
                Precipitation
              </p>
              <p>
                Property: 
                precipitation
              </p>
            </div>
            <div>
              <p>
                Icon: 
                <span>
                  Eye
                </span>
              </p>
              <p>
                Title: 
                Visibility
              </p>
              <p>
                Property: 
                visibility
              </p>
            </div>
          </div>
          <div
            class="emotion-17 emotion-18"
          >
            <div>
              <p>
                Icon: 
                <span>
                  Thermometer
                </span>
              </p>
              <p>
                Title: 
                Dewpoint
              </p>
              <p>
                Property: 
                dewpoint_2m
              </p>
            </div>
            <div>
              <p>
                Icon: 
                <span>
                  Moisture
                </span>
              </p>
              <p>
                Title: 
                Humidity
              </p>
              <p>
                Property: 
                relativehumidity_2m
              </p>
            </div>
          </div>
          <div
            data-testid="HazardLevel"
          >
            <p>
              Hazard: 
              us_aqi
            </p>
          </div>
          <div
            data-testid="HazardLevel"
          >
            <p>
              Hazard: 
              uv_index
            </p>
          </div>
          <div
            data-testid="Wind"
          />
          <div
            data-testid="Pressure"
          />
          <div
            data-testid="SunTime"
          />
        </div>
      </div>
    </div>
  </dialog>
</body>
`;
